drivers/esp8266/sock_udp/udp.c

```diff
@@ -0,0 +1,391 @@
+static sock_udp_ep_t current_ep;
+
+#ifdef MODULE_GNRC_SOCK_CHECK_REUSE
+static sock_udp_t *_udp_socks = NULL;
+#endif
+
+typedef struct transmission {
+    uint8_t *buffer;
+    unsigned int length;
+    sock_udp_ep_t remote;
+} transmission_t;
+
+/* callback functions required by the esp8266 parser drivers/esp8266/esp8266.h */
+void ESP8266_Callback_ClientConnectionConnected(ESP8266_t *ESP8266, ESP8266_Connection_t *Connection)
+{
+    DEBUG("start %s\r\n", __FUNCTION__);
+    DEBUG("Connection name: %s\r\n", Connection->Name);
+    if (0 == strcmp(Connection->Name, TEMP_ID_STR)) {
+        temp_conn = Connection;
+    }
+    if (0 == strcmp(Connection->Name, LOCAL_ID_STR)) {
+        local_conn = Connection;
+    }
+    if (0 == strcmp(Connection->Name, REMOTE_ID_STR)) {
+        remote_conn = Connection;
+    }
+    DEBUG("end %s\r\n", __FUNCTION__);
+}
+void ESP8266_Callback_ClientConnectionError(ESP8266_t *ESP8266, ESP8266_Connection_t *Connection)
+{
+    DEBUG("start %s\r\n", __FUNCTION__);
+    DEBUG("end %s\r\n", __FUNCTION__);
+}
+void ESP8266_Callback_ClientConnectionTimeout(ESP8266_t *ESP8266, ESP8266_Connection_t *Connection)
+{
+    DEBUG("start %s\r\n", __FUNCTION__);
+    DEBUG("end %s\r\n", __FUNCTION__);
+}
+void ESP8266_Callback_ClientConnectionClosed(ESP8266_t *ESP8266, ESP8266_Connection_t *Connection)
+{
+    DEBUG("start %s\r\n", __FUNCTION__);
+    DEBUG("end %s\r\n", __FUNCTION__);
+}
+uint16_t ESP8266_Callback_ClientConnectionSendData(ESP8266_t *ESP8266, ESP8266_Connection_t *Connection, char *Buffer, uint16_t max_buffer_size)
+{
+    DEBUG("start %s\r\n", __FUNCTION__);
+    /* restrain size to avoid memory corruption */
+    if (send_len > max_buffer_size) {
+        send_len = max_buffer_size;
+    }
+    memcpy(Buffer, send_data, send_len);
+    return send_len;
+    DEBUG("end %s\r\n", __FUNCTION__);
+}
+void ESP8266_Callback_ClientConnectionDataSent(ESP8266_t *ESP8266, ESP8266_Connection_t *Connection)
+{
+    // msg_send(NULL, api_pid);
+}
+void ESP8266_Callback_ClientConnectionDataSentError(ESP8266_t *ESP8266, ESP8266_Connection_t *Connection)
+{
+    DEBUG("start %s\r\n", __FUNCTION__);
+    DEBUG("end %s\r\n", __FUNCTION__);
+}
+void ESP8266_Callback_ClientConnectionDataReceived(ESP8266_t *ESP8266, ESP8266_Connection_t *Connection, char *Buffer)
+{
+    DEBUG("start %s\r\n", __FUNCTION__);
+    /* listening endpoint received something */
+    DEBUG("Conection name:  %s\r\n", Connection->Name);
+    if (0 == strcmp(Connection->Name, LOCAL_ID_STR)) {
+        if (sema_wait_timed(&sem_loc_rece_space, INTERVAL_MS * US_PER_MS) < 0) {
+            return;
+        }
+        DEBUG("local endpoint\r\n");
+        /* acutaly, the next line should be
+           rece_len = Connection->BytesReceived;
+           but the program sometimes gives the wrong value when the message ends with a newline */
+        rece_len = Connection->BytesReceived;
+        rece_data = malloc(rece_len);
```

Comment:
Is there really no way around this? Like a global buffer of size [`ETHERNET_DATA_LEN`](https://github.com/RIOT-OS/RIOT/blob/0fcc7d383400db1e3357354f1b94fe3511f78479/sys/include/net/ethernet.h#L34) (minus the fixed header lengths) in this file e.g.?